File: 2/sesion2.hs
import Data.List

-- Retorna una lista de los resultados de 0 hasta n de una funcion f
formaLista f n
  | n == 1 = [f n]
  | otherwise = f n : formaLista f (n - 1)

-- Retorna el factorial de n
fact n
  | n == 0 = 1
  | otherwise = n * fact (n - 1)

-- Retorna el fibonnaci del numero n
fib n
  | n <= 2 = 1
  | otherwise = fib (n - 1) + fib (n - 2)

-- Retorna la funcion de la coyuntura de collatz
collatz :: Fractional a => a -> a
collatz n
  | mod n 2 == 0 = n / 2
  | otherwise = 3 * n - 2

-- Retorna el numero de cifras de un numero
cifras n
  | n < 1 = 0
  | otherwise = 1 + cifras (mod n 10)

-- Retorna  los numeros desde el uno hasta el n en formato 123456789101112
nums n
  | n == 1 = 1
  | otherwise = nums (n - 1) * (10 ^ cifras n) + n

-- Retorna el siguiente intervalo para hallar la raiz de una funcion segun el metodo de biseccion
bisec f lis
  | f ((head lis + head (tail lis)) / 2) * f (head lis) < 0 = [head lis, (head lis + head (tail lis)) / 2]
  | otherwise = [(head lis + head (tail lis)) / 2, head (tail lis)]

-- Itera el metodo de biseccion un dado numero de veces
veces f n lis
  | n == 1 = bisec f lis
  | otherwise = bisec f (veces f (n - 1) lis)

main = do
  print (fact 5)
  print (fib 5)
  print (formaLista fib 5)
  print (collatz 5) --FIXME: Int? Floating?
  print (formaLista collatz 5)
  print (nums 5)
  print (formaLista nums 5)
  print (bisec (\x -> x ^ 2 - 2) [0, 2])
  print (veces (\x -> x ^ 2 - 2) 10 [0, 2])

File: 2/sesion2.py
import sys

# Retorna el factorial de n


def fact(n):
    if (n == 0):
        return 1
    else:
        return n * fact(n-1)

# Retorna la sucesión de fibonacci de longitud n


def fib(n):
    if (n <= 2):
        return 1
    else:
        return fib(n - 1) + fib(n - 2)

# Retorna la lista de las sucesiones de fibonnaci desde 1 hasta n


def listFib(n):
    out = []
    for i in range(n):
        out.append(fib(i + 1))
    return out

# Retorna función de la coyuntura de collatz


def collatz(n):
    if (n % 2 == 0):
        return n / 2
    else:
        return 3 * n + 1

# Retorna una lista con las coyunturas de collatz desde 0 hasta n


def listColl(n):
    out = []
    for i in range(n + 1):
        out.append(collatz(i))
    return out

# Retorna el numero de cifras en un dado numero


def cifras(n):
    c = 0
    while (n > 0):
        n = n // 10
        c += 1
    return c

# Retorna los numeros desde el 1 hasta n en formato: 123456789101112


def nums(n):
    if (n == 1):
        return 1
    else:
        return nums(n - 1) * (10 ** (cifras(n))) + n

# Retorna una lista con los numeros mostrados con el formato anterior desde el 1 hasta el n


def listNums(n):
    out = []
    for i in range(n):
        out.append(nums(i + 1))
    return out

# Funcion


def f(x):
    return x**2 - 2

# Siguiente intervalo para hallar la raiz de una funcion segun el metodo de biseccion


def bisec(lis):
    if f((lis[0] + lis[1]) / 2) * f(lis[0]) < 0:
        return [lis[0], (lis[1] + lis[0]) / 2]
    else:
        return [(lis[1] + lis[0]) / 2, lis[1]]

# Itera el metodo de biseccion un dado numero de veces


def veces(n):
    if (n == 1):
        return bisec([0, 2])
    else:
        return bisec(veces(n - 1))

# MAIN


def main(argv):
    print(argv + "! = " + str(fact(int(argv))))
    print("fib (" + argv + ") = " + str(fib(int(argv))))
    print("listFib (" + argv + ") = " + str(listFib(int(argv))))
    print("collatz (" + argv + ") = " + str(collatz(int(argv))))
    print("listColl (" + argv + ") = " + str(listColl(int(argv))))
    print("nums (" + argv + ") = " + str(nums(int(argv))))
    print("listNums (" + argv + ") = " + str(listNums(int(argv))))
    print("bisec([0, 2]) = " + str(bisec([0, 2])))
    print("veces (10) = " + str(veces(10)))


if __name__ == '__main__':
    if (len(sys.argv) == 2):
        main(sys.argv[1])
    else:
        print("ERROR: Se debe introducir:\n\tsesion1.py <numero>")
